// qs.js/src/polyfill/list.js
function remove(arr, index) {
    let originLen = arr.length, result;
    if (index >= 0 && index < originLen) {
        if (originLen === 1) {
            result = [];
        } else {
            result = new Array(originLen - 1);
            for (let i = 0, j = 0; i < originLen; i++) {
                if (i !== index) {
                    result[j] = arr[i];
                    j++;
                }
            }
        }
    } else {
        result = arr;
    }
    return result;
}


function concat(a, b) {
    return a.concat(b);
}


function find(arr, v) {
    for (let i = 0, len =arr.length; i < len; i++) {
        if (arr[i] === v) {
            return i;
        }
    }
    return -1;
}

function findBy(arr, pred) {
    for (let i = 0, len =arr.length; i < len; i++) {
        if (pred(arr[i])) {
            return i;
        }
    }
    return -1;
}

function contains(arr, val) {
    return find(arr, val) !== -1;
}


function forEach(arr, fn) {
    for (let i = 0 , len = arr.length; i < len; i++) {
        fn(arr[i], i);
    }
}

function map(arr, fn) {
    let len = arr.length;
    var mapArr = new Array(len);
    for (let i = 0; i < len; i++) {
        mapArr[i] = fn(arr[i], i);
    }
    return mapArr;
}

function fill(arr, val) {
    var len = arr.length,
        i;
    for (i = 0; i < len; i++) {
        arr[i] = val;
    }
}

export {remove, concat, find, findBy, contains, forEach, fill, map};
